### PR TITLE
 dev,v3.0/benchmark: add 'add-index-with-load.md'

### PR DESCRIPTION
**Pull request by bb7133**
### What is changed, added or deleted? 

Add a new benchmark result on `ALTER TABLE ... ADD INDEX` with online workload

### What is the related PR or file link(s)? 
N/A

### Which version does your change affect? 
dev and v3.0

### CODE REVIEW

**Comment by bb7133** [comment]
hi @lilin90 @yikeke  PTAL, thanks

File: dev/benchmark/add-index-with-load.md [hunk 1]
```diff
@@ -0,0 +1,345 @@
+---
```
Comment (lilin90) [inline dev/benchmark/add-index-with-load.md]: Add this file to `TOC.md`?

File: dev/benchmark/add-index-with-load.md [hunk 2]
```diff
@@ -0,0 +1,345 @@
...
+title: 线上负载与 Add Index 相互影响测试
+category: benchmark
+aliases: ['/docs-cn/benchmark/add-index-with-load.md/']
```
Comment (lilin90) [inline dev/benchmark/add-index-with-load.md]: `aliases` is unnecessary for a newly added file. Please remove it.

File: dev/benchmark/add-index-with-load.md [hunk 3]
```diff
@@ -0,0 +1,345 @@
...
+---
+
+# 线上负载与 Add Index 相互影响测试
+
+## 测试目的
+
+对比 OLTP 场景下，Add Index 对与线上负载相互的影响
```
Comment (lilin90) [inline dev/benchmark/add-index-with-load.md]:
```suggestion
对比 OLTP 场景下，`Add Index` 与线上负载的相互影响。
```

File: dev/benchmark/add-index-with-load.md [hunk 4]
```diff
@@ -0,0 +1,345 @@
...
+
+TiDB 版本：v3.0.1
+
+时间：2019 年 7 月
+
+地点：北京
+
+## 测试环境
+
+3 TiDB，3 TiKV，3 PD on K8s cluster。
```
Comment (lilin90) [inline dev/benchmark/add-index-with-load.md]:
```suggestion
Kubernetes 集群上 3 个 TiDB 实例，3 个 TiKV 实例，3 个 PD 实例。
```

File: dev/benchmark/add-index-with-load.md [hunk 5]
```diff
@@ -0,0 +1,345 @@
...
+| :--- | :---------------------------------------- |
+| TiDB  | `9e4e8da3c58c65123db5f26409759fe1847529f8` |
+| TiKV  | `4151dc8878985df191b47851d67ca21365396133` |
+|  PD   | `811ce0b9a1335d1b2a049fd97ef9e186f1c9efc1` |
+
+Sysbench 版本：1.0.17
+
+### TiDB 参数配置
+
+TiKD / TiKV / PD 均使用 [TiDB-Operator](https://github.com/pingcap/tidb-operator) 默认配置。
```
Comment (lilin90) [inline dev/benchmark/add-index-with-load.md]:
```suggestion
TiKD/TiKV/PD 均使用 [TiDB Operator](https://github.com/pingcap/tidb-operator) 默认配置。
```

File: dev/benchmark/add-index-with-load.md [hunk 6]
```diff
@@ -0,0 +1,345 @@
...
+    --rand-seed=$RANDOM \
+    --db-driver=mysql \
+    --mysql-db=sbtest \
+    --mysql-host=$tidb_host \
+    --mysql-port=$tidb_port \
+    --mysql-user=root \
+    run --tables=1 --table-size=2000000
+```
+
+## 测试方案1 -- Add Index 目标列被频繁 Update
```
Comment (lilin90) [inline dev/benchmark/add-index-with-load.md]:
```suggestion
## 测试方案 1：Add Index 目标列被频繁 Update
```

File: dev/benchmark/add-index-with-load.md [hunk 7]
```diff
@@ -0,0 +1,345 @@
...
+
+1. 开始 `oltp_read_write` 测试。
+2. 与步骤 1 同时，使用 `alter table sbtest1 add index c_idx(c)` 添加索引。
+3. 在步骤 2 结束，即索引添加完成时，停止步骤 1 的测试。
+4. 获取指标：`alter table ... add index` 的运行时间；sysbench 在该时间段内的平均 TPS 和 QPS。
```
Comment (lilin90) [inline dev/benchmark/add-index-with-load.md]:
```suggestion
4. 获取指标 `alter table ... add index` 的运行时间；sysbench 在该时间段内的平均 TPS 和 QPS。
```

File: dev/benchmark/add-index-with-load.md [hunk 8]
```diff
@@ -0,0 +1,345 @@
...
+5. 逐渐增大 `tidb_ddl_reorg_worker_cnt` 和 `tidb_ddl_reorg_batch_size` 两个参数，重复步骤 1-4。
```
Comment (lilin90) [inline dev/benchmark/add-index-with-load.md]:
```suggestion
5. 逐渐增大 `tidb_ddl_reorg_worker_cnt` 和 `tidb_ddl_reorg_batch_size` 两个参数的值，重复步骤 1-4。
```

File: dev/benchmark/add-index-with-load.md [hunk 9]
```diff
@@ -0,0 +1,345 @@
...
+| :------------------------ | :---------------------- | :------------- | :----------- |
+| 1     | 402 |  338.4 |           6776 |
+| 2     | 266 |  330.3 |           6001 |
+| 4     | 174 |  288.5 |           5769 |
+| 8     | 129 |  280.6 |           5612 |
+| 16    | 90 |   263.5 |           5273 |
+| 32    | 54 |   229.2 |           4583 |
+| 48    | 57 |   230.1 |           4601 |
+
+![add-index-load-1-b32](../../media/add-index-load-1-b32.png)
```
Comment (lilin90) [inline dev/benchmark/add-index-with-load.md]:
```suggestion
![add-index-load-1-b32](/media/add-index-load-1-b32.png)
```

File: dev/benchmark/add-index-with-load.md [hunk 10]
```diff
@@ -0,0 +1,345 @@
...
+| :------------------------ | :---------------------- | :------------- | :----------- |
+| 1     | 264 |  269.4 |           5388 |
+| 2     | 163 |  266.2 |           5324 |
+| 4     | 105 |  272.5 |           5430 |
+| 8     | 78 |  262.5 |           5228 |
+| 16    | 57 |   215.5 |           4308 |
+| 32    | 42 |   185.2 |           3715 |
+| 48    | 45 |   189.2 |           3794 |
+
+![add-index-load-1-b64](../../media/add-index-load-1-b64.png)
```
Comment (lilin90) [inline dev/benchmark/add-index-with-load.md]:
```suggestion
![add-index-load-1-b64](/media/add-index-load-1-b64.png)
```

File: dev/benchmark/add-index-with-load.md [hunk 11]
```diff
@@ -0,0 +1,345 @@
...
+| :------------------------ | :---------------------- | :------------- | :----------- |
+| 1     | 171 |  289.1 |           5779 |
+| 2     | 110 |  274.2 |           5485 |
+| 4     | 79 |  250.6 |           5011 |
+| 8     | 51 |  246.1 |           4922 |
+| 16    | 39 |   171.1 |           3431 |
+| 32    | 35 |   130.8 |           2629 |
+| 48    | 35 |   120.5 |           2425 |
+
+![add-index-load-1-b128](../../media/add-index-load-1-b128.png)
```
Comment (lilin90) [inline dev/benchmark/add-index-with-load.md]:
```suggestion
![add-index-load-1-b128](/media/add-index-load-1-b128.png)
```

File: dev/benchmark/add-index-with-load.md [hunk 12]
```diff
@@ -0,0 +1,345 @@
...
+| :------------------------ | :---------------------- | :------------- | :----------- |
+| 1     | 145 |  283.0 |           5659 |
+| 2     | 96 |  282.2 |           5593 |
+| 4     | 56 |  236.5 |           4735 |
+| 8     | 45 |  194.2 |           3882 |
+| 16    | 39 |   149.3 |           2893 |
+| 32    | 36 |   113.5 |           2268 |
+| 48    | 33 |   86.2 |           1715 |
+
+![add-index-load-1-b256](../../media/add-index-load-1-b256.png)
```
Comment (lilin90) [inline dev/benchmark/add-index-with-load.md]:
```suggestion
![add-index-load-1-b256](/media/add-index-load-1-b256.png)
```

File: dev/benchmark/add-index-with-load.md [hunk 13]
```diff
@@ -0,0 +1,345 @@
...
+| :------------------------ | :---------------------- | :------------- | :----------- |
+| 1     | 135 |  257.8 |           5147 |
+| 2     | 78 |  252.8 |           5053 |
+| 4     | 49 |  222.7 |           4478 |
+| 8     | 36 |  145.4 |           2904 |
+| 16    | 33 |   109 |           2190 |
+| 32    | 33 |   72.5 |           1503 |
+| 48    | 33 |   54.2 |           1318 |
+
+![add-index-load-1-b512](../../media/add-index-load-1-b512.png)
```
Comment (lilin90) [inline dev/benchmark/add-index-with-load.md]:
```suggestion
![add-index-load-1-b512](/media/add-index-load-1-b512.png)
```

File: dev/benchmark/add-index-with-load.md [hunk 14]
```diff
@@ -0,0 +1,345 @@
...
+| :------------------------ | :---------------------- | :------------- | :----------- |
+| 1     | 111 |  244.3 |           4885 |
+| 2     | 78 |  228.4 |           4573 |
+| 4     | 54 |  168.8 |           3320 |
+| 8     | 39 |  123.8 |           2475 |
+| 16    | 36 |   59.6 |           1213 |
+| 32    | 42 |   93.2 |           1835 |
+| 48    | 51 |   115.7 |           2261 |
+
+![add-index-load-1-b1024](../../media/add-index-load-1-b1024.png)
```
Comment (lilin90) [inline dev/benchmark/add-index-with-load.md]:
```suggestion
![add-index-load-1-b1024](/media/add-index-load-1-b1024.png)
```

File: dev/benchmark/add-index-with-load.md [hunk 15]
```diff
@@ -0,0 +1,345 @@
...
+| :------------------------ | :---------------------- | :------------- | :----------- |
+| 1     | 918 |  243.3 |           4855 |
+| 2     | 1160 |  209.9 |           4194 |
+| 4     | 342 |  185.4 |           3707 |
+| 8     | 1316 |  151.0 |           3027 |
+| 16    | 795 |   30.5 |           679 |
+| 32    | 1130 |   26.69 |           547 |
+| 48    | 893 |   27.5 |           552 |
+
+![add-index-load-1-b2048](../../media/add-index-load-1-b2048.png)
```
Comment (lilin90) [inline dev/benchmark/add-index-with-load.md]:
```suggestion
![add-index-load-1-b2048](/media/add-index-load-1-b2048.png)
```

File: dev/benchmark/add-index-with-load.md [hunk 16]
```diff
@@ -0,0 +1,345 @@
...
+| :------------------------ | :---------------------- | :------------- | :----------- |
+| 1     | 3042 |  200.0 |           4001 |
+| 2     | 3022 |  203.8 |           4076 |
+| 4     | 858 |  195.5 |            3971 |
+| 8     | 3015 |  177.1 |           3522 |
+| 16    | 837 |   143.8 |           2875 |
+| 32    | 942 |   114 |           2267 |
+| 48    | 187 |   54.2 |           1416 |
+
+![add-index-load-1-b4096](../../media/add-index-load-1-b4096.png)
```
Comment (lilin90) [inline dev/benchmark/add-index-with-load.md]:
```suggestion
![add-index-load-1-b4096](/media/add-index-load-1-b4096.png)
```

File: dev/benchmark/add-index-with-load.md [hunk 17]
```diff
@@ -0,0 +1,345 @@
...
+
+### 测试结论
+
+若 Add Index 的目标列正在进行较为频繁的写操作（本测试中涉及列的 update/insert/delete），默认 Add Index 配置对系统的线上负载有比较明显的影响，该影响主要来源于 Add Index 与 Column Update 并发进行造成的写冲突，系统的表现反应在：
+
+- 随着两个参数的逐渐增大，TiKV prewrite latch wait duration 有明显的升高，造成写入变慢。
+- `tidb_ddl_reorg_worker_cnt` 与 `tidb_ddl_reorg_batch_size` 非常大时，`admin show ddl` 命令可以看到 ddl job 的多次重试（例如 `Write conflict, txnStartTS 410327455965380624 is stale [try again later], ErrCount:38, SnapshotVersion:410327228136030220`），此时 Add Index 会持久非常久才能完成。
```
Comment (lilin90) [inline dev/benchmark/add-index-with-load.md]:
```suggestion
- `tidb_ddl_reorg_worker_cnt` 与 `tidb_ddl_reorg_batch_size` 非常大时，`admin show ddl` 命令可以看到 ddl job 的多次重试（例如 `Write conflict, txnStartTS 410327455965380624 is stale [try again later], ErrCount:38, SnapshotVersion:410327228136030220`），此时 Add Index 会持续非常久才能完成。
```

File: dev/benchmark/add-index-with-load.md [hunk 18]
```diff
@@ -0,0 +1,345 @@
...
+
+## 测试方案2 -- Add Index 目标列不涉及写入（仅查询）
```
Comment (lilin90) [inline dev/benchmark/add-index-with-load.md]:
```suggestion
## 测试方案2：Add Index 目标列不涉及写入（仅查询）
```

File: dev/benchmark/add-index-with-load.md [hunk 19]
```diff
@@ -0,0 +1,345 @@
...
+
+1. 开始 `oltp_read_only` 测试。
+2. 与步骤 1 同时，使用 `alter table sbtest1 add index c_idx(c)` 添加索引。
+3. 在步骤 2 结束，即索引添加完成时，停止步骤 1。
+4. 获取指标：`alter table ... add index` 的运行时间；sysbench 在该时间段内的平均 TPS 和 QPS。
```
Comment (lilin90) [inline dev/benchmark/add-index-with-load.md]:
```suggestion
4. 获取指标 `alter table ... add index` 的运行时间；sysbench 在该时间段内的平均 TPS 和 QPS。
```

File: dev/benchmark/add-index-with-load.md [hunk 20]
```diff
@@ -0,0 +1,345 @@
...
+| :------------------------ | :---------------------- | :------------- | :----------- |
+| 1     | 376 |  548.9 |           8780 |
+| 2     | 212 |  541.5 |           8523 |
+| 4     | 135 |  538.6 |           8549 |
+| 8     | 114 |  536.7 |           8393 |
+| 16    | 77 |   533.9 |           8292 |
+| 32    | 46 |   533.4 |           8103 |
+| 48    | 46 |   532.2 |           8074 |
+
+![add-index-load-2-b32](../../media/add-index-load-2-b32.png)
```
Comment (lilin90) [inline dev/benchmark/add-index-with-load.md]: Please update the path for all images. This is to guarantee they can display well on the PingCAP website.

File: dev/benchmark/add-index-with-load.md [hunk 21]
```diff
@@ -0,0 +1,345 @@
...
+### 测试结论
+
+对于 Add Index 的目标列仅有查询负载时，Add Index 对负载的影响不明显。
+
+## 测试方案3 -- 集群负载不涉及 Add Index 目标列
+
+1. 开始 `oltp_read_write` 测试。
+2. 与步骤 1 同时，使用 `alter table test add index pad_idx(pad)` 添加索引。
+3. 在步骤 2 结束，即索引添加完成时，停止步骤 1 的测试。
+4. 获取指标：`alter table ... add index` 的运行时间，sysbench 在该时间段内的平均 TPS 和 QPS。
```
Comment (lilin90) [inline dev/benchmark/add-index-with-load.md]:
```suggestion
4. 获取指标 `alter table ... add index` 的运行时间，sysbench 在该时间段内的平均 TPS 和 QPS。
```

File: dev/benchmark/add-index-with-load.md [hunk 22]
```diff
@@ -0,0 +1,345 @@
...
+| 8     | 37 |  279.5 |           5612 |
+| 16    | 34 |   250.4   |           5365 |
+| 32    | 32 |   220.2 |           4924 |
+| 48    | 33 |   214.8 |           4544 |
+
+![add-index-load-3-b4096](../../media/add-index-load-3-b4096.png)
+
+### 测试结论
+
+对于 Add Index 的目标列与负载无关时，Add Index 对负载的影响不明显。
```
Comment (lilin90) [inline dev/benchmark/add-index-with-load.md]:
```suggestion
对于 `Add Index` 的目标列与负载无关时，`Add Index` 对负载的影响不明显。
```

File: dev/benchmark/add-index-with-load.md (hunk 20 above)
Comment (bb7133) [reply dev/benchmark/add-index-with-load.md]: Done, thanks

**Comment by bb7133** [comment]
hi @lilin90 , thank you very much for all the detailed comments. I've addressed them.

**Pull request merged**